libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2132,55 +2118,93 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
 	// For positional calls only, store argument types
 	if (_functionCall.names().empty())
 	{
-		shared_ptr<TypePointers> const argumentTypes = make_shared<TypePointers>();
+		shared_ptr<TypePointers> argumentTypes = make_shared<TypePointers>();
 		for (ASTPointer<Expression const> const& argument: arguments)
 			argumentTypes->push_back(type(*argument));
 		_functionCall.expression().annotation().argumentTypes = move(argumentTypes);
 	}
 
 	_functionCall.expression().accept(*this);
+
 	TypePointer const expressionType = type(_functionCall.expression());
+	TypePointer actualType;
 
+	// Determine type for type conversion or struct construction expressions
 	if (auto const* const typeType = dynamic_cast<TypeType const*>(expressionType.get()))
+		actualType = typeType->actualType();
+
+	// Determine annotation kind for this FunctionCall node
+	const FunctionCallKind annoKind =
+		!actualType
+			? FunctionCallKind::FunctionCall
+		: actualType->category() == Type::Category::Struct
+			? FunctionCallKind::StructConstructorCall
+			: FunctionCallKind::TypeConversion;
+
+	// Determine function type for function call or struct construction expressions
+	FunctionTypePointer const functionType =
+		annoKind == FunctionCallKind::StructConstructorCall ?
+			dynamic_cast<StructType const&>(*actualType).constructorType() :
+		annoKind == FunctionCallKind::FunctionCall ?
+			dynamic_pointer_cast<FunctionType const>(expressionType) :
+			nullptr;
+
+	// Determine purity; for function calls, this also depends upon the callee and its FunctionType
+	const bool annoPure = functionType && annoKind == FunctionCallKind::FunctionCall
```

Comment:
Do you really need this intermediate variable?

Reply:
No; removed.